New connection page design

Jamakase opened the pull request:
## What
Includes a bunch of improvements for Connection Form:

- Batch Edit
![image](https://user-images.githubusercontent.com/3767150/153867262-289ff20c-9494-44ce-8470-87dea6e7a817.png)

- New Connection Form UI 
  - New StreamFields table 
![image](https://user-images.githubusercontent.com/3767150/153868521-0c9ab215-1a55-4cd2-823e-ecc6ba3ae322.png)
  - New tooltips for table headers ![image](https://user-images.githubusercontent.com/3767150/153867174-e0a2f33c-e796-46ec-9e79-0fb06bf355d3.png)
  - Update stream meta info 
![image](https://user-images.githubusercontent.com/3767150/153868601-01539507-ce2a-4fc8-8df3-b4b53a5bc824.png)
  - Add Collapsible Cards 
![image](https://user-images.githubusercontent.com/3767150/153868641-a1f28bfe-1a5c-48a8-b9b1-73a2dc3fe05a.png)
  - New settings navigation
![image](https://user-images.githubusercontent.com/3767150/153868686-ff2563d6-4298-466e-800f-1f5c509ea9af.png)
- Move transformation to separate tab 
![image](https://user-images.githubusercontent.com/3767150/153478145-ab3fe01a-ddbf-48c9-9859-5c5929f66a77.png)
- Set 24h sync by default Closes #9733


Closes #9728 
Closes #9729
Closes #4226
Closes #9462 

What is left:
- [ ] Search section for primary key
- [ ] Search in `PathPopout` ( primary key and cursor field )
- [ ] Update onboarding
- [ ] Migrate table to react-table ( for better support of tables and proper column resizes )
- [ ] Minor designs update ( checkbox positioning, paddings in table )

## Recommended reading order
1. `ConnectionForm.tsx`
2. `CollapsibleCard.tsx`
3. `FormCard.tsx`
4. `services/BulkEdit/BulkEditService.tsx`
5. `BulkEditHeader.tsx`
6. `PathPopout.tsx`

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/airbytehq/airbyte/9025)
<!-- Reviewable:end -->

jrhizor commented:
Is this ready for me to take a look at?

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/airbytehq/airbyte?pullRequest=9025) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/airbytehq/airbyte?pullRequest=9025) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: Jamakase<br/>:x: juliachvyrova<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/airbytehq/airbyte?pullRequest=9025) it.</sub>

timroes commented:
* We should have a message if the Transformation section is empty, stating that this connection doesn't support any transformations.
* :bug: When the "Replication frequency" is set to "Manual" it will now always clear the select to an empty "..." instead of showing manual.
* Since most of the settings are now moved to the "Replication" and "Transformation" page, let's change the "gear icon" link in the connections overview table to lead to the replication page instead of the settings page, which now only holds the delete button.
* We should rename the Settings tab to something more fitting now, but I currently don't have a good idea what would be a better name, now since it only contains the Delete button. Will reach out to some people getting feedback on this, but any ideas welcome.

timroes commented:
With this change we now show the "Don't worry! You'll be able to change this later on" below the stream configuration directly:

![screenshot-20220216-130240](https://user-images.githubusercontent.com/877229/154260918-23d9d76a-0709-462d-bd7f-3b9094345f0c.png)

Since actually changing the streams later, will require the data to be completely reset and fetched again, I am not sure if it's still worth showing this hint at all, since it might make it sound like "Don't think about it yet, it's easily configurable later", but it would infact later affect your data changing this.

@misteryeo you've some thoughts around this?

timroes commented:
Before this PR the destination namespace was set to "Mirror source structure" by default, now it's unselected by default when creating a connection. Was this a deliberate change?

Jamakase commented:
> Before this PR the destination namespace was set to "Mirror source structure" by default, now it's unselected by default when creating a connection. Was this a deliberate change?

Yep. This is a fix for https://github.com/airbytehq/airbyte/issues/9728

misteryeo commented:
> Since actually changing the streams later, will require the data to be completely reset and fetched again, I am not sure if it's still worth showing this hint at all, since it might make it sound like "Don't think about it yet, it's easily configurable later", but it would in fact later affect your data changing this.

Has this always been the case? When you update anything in the stream, has it always required a complete reset?

It feels like this behavior is something we want to improve in the backend otherwise we'd want almost a warning that tells them what they setup now can't be changed without the data being completely reset and fetched again - which is actually more concerning.

Jamakase commented on airbyte-webapp/src/views/Connection/CatalogTree/components/PathPopout.tsx:
```diff
@@ -0,0 +1,92 @@
+    <Popout
+      options={options}
+      value={props.path}
+      // @ts-ignore need to solve issue with typings
```
Actually, I think we should first properly set up linting if we are going to take care of such cases, so we won't leave comments for every wrong used annotation.

Jamakase commented on airbyte-webapp/src/views/Connection/ConnectionForm/formConfig.tsx:
```diff
@@ -240,64 +230,76 @@ const useInitialSchema = (schema: SyncSchema): SyncSchema =>
+  isEditMode?: boolean
+): NormalizationType => {
+  let initialNormalization = (operations.find(
+    (op) => op.operatorConfiguration.operatorType === OperatorType.Normalization
```
Wow. Awesome idea! Good catch @timroes.

Jamakase commented on airbyte-webapp/src/views/Connection/ConnectionForm/DefaultSyncSettingsField.tsx:
```diff
@@ -0,0 +1,136 @@
+};
+
+// TODO: disabled for now. Need to clarify that these fields are not duplicated
+const DefaultSyncSettingsField: React.FC<DropdownProps> = (props) => {
```
I think we won't need it. I put it back so it won't be lost.

Jamakase commented on airbyte-webapp/src/locales/en.json:
```diff
@@ -125,14 +124,24 @@
 
-  "form.rawData": "Raw data - no normalization",
-  "form.basicNormalization": "Basic normalization - ",
+  "form.rawData": "Serialized JSON",
+  "form.basicNormalization": "Original structure",
```
I believe I may indeed use the wrong names for the wrong values. So `Serialized JSON` is for `basicNormalization` right?

timroes replied:
That was my understanding. @misteryeo could you maybe have a look at that. And also would like to hear your thoughts around "Serialized JSON" vs something like "Flattened JSON" or "Normalized JSON"?

misteryeo replied:
Yeah, I think the name and values are switched here.

I don't have an opinion on the use of "Serialized JSON" - I've not seen or heard any confusion around this naming previously though either on blogs mentioning Airbyte or in our Slack community. Conversely, I've not seen anyone refer to flatted or normalized JSON. @ChristopheDuong since you've been working on normalization, do you have thoughts on this?

Jamakase replied:
Switched values

ChristopheDuong replied:
"serialized json" sounds like the string representation of a json object, so it would be the raw data instead (the way it was first before switched)?

So this labelling does sound confusing...

jrhizor replied:
@ChristopheDuong what do you recommend for labelling? Should we just keep the old version?

misteryeo replied:
What was the original labeling?

ChristopheDuong replied:
maybe normalized records should not denote anything about "json" (which is the internal representation of records/messages in airbyte protocol). 

Either we keep the old naming or a variant (plus if we are changing the naming completely here, we may have to update a lot of docs too?)

What about `"Normalized Tabular data"`?

and for the raw data, you can name "Serialized JSON blob", or "json blob", "json format"?

jrhizor replied:
The originals were "Raw data - no normalization" and "Basic normalization".

The raw one does store data as raw JSON. Given that the new view labels it as normalization options, it may make sense to have "Raw data (JSON)" and "Split into separate tables" or "Normalized tabular data" like Chris suggested.

Jamakase commented on airbyte-webapp/src/views/Connection/ConnectionForm/formConfig.tsx:
```diff
@@ -240,64 +230,76 @@ const useInitialSchema = (schema: SyncSchema): SyncSchema =>
 
+const getInitialTransformations = (operations: Operation[]): Transformation[] =>
+  (operations.filter(
+    (op) => op.operatorConfiguration.operatorType === OperatorType.Dbt
```
Hmm `?? []` looks a bit strange. Probably typings are wrong somewhere. I need to validate a few cases and make sure that we do not need this condition.

Jamakase replied:
Looks like this could be indeed removed.

Jamakase commented on airbyte-webapp/src/views/Connection/ConnectionForm/ConnectionForm.tsx:
```diff
@@ -11,44 +12,59 @@ import { ControlLabels, DropDown, DropDownRow, Input, Label } from "components";
+  justify-content: flex-start;
+  align-items: flex-start;
+
+  & > div:not(:last-child) {
+    margin-right: 10px;
+  }
```
Nope. I am just not that good at grids and use everything in an old-fashioned way. Thank you for the tip

Jamakase commented:
> * 🐛 When the "Replication frequency" is set to "Manual" it will now always clear the select to an empty "..." instead of showing manual.
> * Since most of the settings are now moved to the "Replication" and "Transformation" page, let's change the "gear icon" link in the connections overview table to lead to the replication page instead of the settings page, which now only holds the delete button.

Addressed.

> * We should have a message if the Transformation section is empty, stating that this connection doesn't support any transformations.
![image](https://user-images.githubusercontent.com/3767150/154562214-9cd54aef-172a-4acd-9a76-5d3629835638.png)

Jamakase commented on airbyte-webapp/src/views/Connection/ConnectionForm/components/InformationToolTip.tsx:
```diff
@@ -0,0 +1,38 @@
+import ToolTip, { InfoIcon } from "components/ToolTip";
+
+const ToolTipBlock = styled(ToolTip)`
+  top: calc(100% + 0px);
```
hmm, nope. Looks like I just copy-pasted it from Tooltip and forgot to remove.

Jamakase commented on airbyte-webapp/src/views/Connection/ConnectionForm/components/SyncCatalogField.tsx:
```diff
@@ -1,64 +1,198 @@
-const RadioButtonControl = styled(LabeledRadioButton)`
-  margin: 0 0 0 5px;
+const LearnMoreLink = styled.a`
+  opacity: 0.5;
```
But this color is picked from figma. Should we still change it to 0.6?

timroes replied:
Yes, please change it.

Jamakase replied:
Done

Jamakase commented on airbyte-webapp/src/views/Connection/CatalogTree/components/BulkHeader.tsx:
```diff
@@ -0,0 +1,176 @@
+  const cursorRequired = options.syncMode === SyncMode.Incremental;
+  // const shouldDefinePk =
+  //   stream.sourceDefinedPrimaryKey.length === 0 && pkRequired;
+  // const shouldDefineCursor = !stream.sourceDefinedCursor && cursorRequired;
```
Updated this part to also take into consideration sourceDefined options

Jamakase commented on airbyte-webapp/src/views/Connection/ConnectionForm/TransferFormCard.tsx:
```diff
@@ -0,0 +1,96 @@
+              options={frequencies}
+              onChange={(item) => {
+                form.setFieldValue(field.name, item.value);
+                // reset();
```
Removed this part

timroes reviewed (approved):
Tested this. Functionality wise seems to work. I've mostly have minor concerns around design and code left, but we should not block this PR now on this, and fix them afterwards. So LGTM for me to merge.

Jamakase commented on airbyte-webapp/src/components/base/DropDown/components/Option.tsx:
```diff
@@ -45,10 +45,16 @@ export const OptionView = styled.div<{
 const Option: React.FC<IProps> = (props) => {
+  const dataTestId = props.data.testId
+    ? props.data.testId
+    : !["object", "array"].includes(typeof props.data.label)
+    ? props.data.label
+    : `select_value_${props.data.value}`;
```
That's exactly the reason why I added this logic. By default, we were setting data-testid to label, so if we change it right now to value - it will break some e2e tests.

Should we open separate task for it? Or should I fix it as part of this PR?

timroes replied:
Let's push it into a separate PR.

timroes commented on airbyte-webapp/src/locales/en.json:
```diff
@@ -87,11 +84,13 @@
-  "form.prefix.placeholder": "prefix-",
-  "form.nameSearch": "Search name",
+  "form.prefix": "Destination Stream Prefix (Optional)",
+  "form.prefix.message": "Add a prefix to stream names (ex. “airbyte” causes “projects” => “airbyte_projects”)",
```
Let's leave it for now. And I'll follow up with Jared around that this week.
